### PR TITLE
Fix unsafe hotserving behaviour for non-multimedia uploads.

### PR DESCRIPTION
**Pull request by joshqou**
Ensures that the Media API only returns a disposition type of `inline` for media, return `attachment` for everything else. NVT#1548992

Signed-off-by: Josh Qou <jqou@icloud.com>

### CODE REVIEW

**Comment by joshqou** [comment]
Patch works but synapse has some strict tests that are failing, can't find where the tests referencing perl are so not much I can do.

**Comment by clokep** [comment]
> Patch works but synapse has some strict tests that are failing, can't find where the tests referencing perl are so not much I can do.

It looks like linting is failing, you should be able to run `black` (or `scripts-dev/lint.sh`) and it should fix it for you.

**Comment by joshqou** [comment]
@clokep The actual fails relate to the media api tests, it's just the change I tried making that's causing the lint fail. I'm not too familiar with poetry so I'd appreciate some help from a more experienced synapse contributor with fixing the tests so this can get merged in.

File: synapse/media/_base.py [hunk 1]
```diff
@@ -151,7 +153,16 @@ def _quote(x: str) -> str:
     else:
         content_type = media_type
 
+    # Only hotserve "safe" mimetypes, force download everything else
+    disposition_type = "attachment"
+    for mime in HOTSERVE_CONTENT_TYPES:
+        if media_type.lower().startswith(mime):
+            disposition_type = "inline"
+            break
```
Comment (clokep) [inline synapse/media/_base.py:161]: From chatting internally it seems safer to just always use an `attachment` content disposition. There isn't really any benefit to using `inline` as things like `` tags ignore the `Content-Disposition` header anyway.

Comment (joshqou) [reply synapse/media/_base.py:161]: If `` and `<video>` ignore inline that seems sensible

File: synapse/media/_base.py [hunk 2]
```diff
@@ -151,7 +151,9 @@ def _quote(x: str) -> str:
     else:
         content_type = media_type
 
+    disposition = "attachment"
     request.setHeader(b"Content-Type", content_type.encode("UTF-8"))
+
     if upload_name:
```
Comment (clokep) [inline synapse/media/_base.py:157]:
```suggestion
    request.setHeader(b"Content-Type", content_type.encode("UTF-8"))

    # Use a Content-Disposition of attachment to force download of media.
    disposition = "attachment"
    if upload_name:
```

**Comment by erikjohnston** [comment]
Does this also force downloading of things like PDFs? I know that users do appreciate the ability to click on the PDF and have it open in a browser tab.

**Comment by reivilibre** [comment]
> Does this also force downloading of things like PDFs? I know that users do appreciate the ability to click on the PDF and have it open in a browser tab.

I tested this in Firefox and Element Web.

What I observed:

* both before and after this change, if you click on a PDF attachment (in an unencrypted room) in Element Web, the PDF is downloaded **and** opened in a browser tab (using default settings for PDF handling in Firefox) — the downloaded copy is what is opened (via `file:///` URI).
* before this change, accessing a PDF file directly `https://syn1:8448/_matrix/media/v3/download/syn1/DCwMApqkfYSGtJSmpiJgXkiw` would open it inline
* after this change, accessing a PDF file directly as above does the same as what Element Web does: the PDF is downloaded and Firefox opens the PDF in a new tab via `file:///`.

What's crucially notable is that Element Web was unaffected.
I am ignoring the IRC bridging concern because I'm not familiar, but I guess that would hit the difference noted in points 2 and 3.

Given the security benefit, I don't see anything that should put us off here.

**Comment by reivilibre** [comment]
The only things I can see blocking us from taking this right now:

- Complement (https://github.com/matrix-org/complement) is failing, it probably needs a bit of `s/inline/attachment/` in a few places
- Sytest is failing unless we apply your PR (https://github.com/matrix-org/sytest/pull/1353) — thanks — but if we apply that PR then Dendrite's CI will fail and it would be nice to give them some warning or fix it up for them if easy to do

Other than that I think it would be fine?

**Comment by joshqou** [comment]
@reivilibre Made PRs for dendrite and complement.
- https://github.com/matrix-org/dendrite/pull/3113
- https://github.com/matrix-org/complement/pull/639

**Pull request merged**